[Bug]: Merge rows (diff) + ETL metadata injection - possibly breaking change

**[UPDATE: a detailed use case regarding this issue is explained in this post below](https://github.com/apache/hop/issues/6069#issuecomment-3575285863)**

### Apache Hop version?

2.16

### Java version?

17.0.2

### Operating system

Linux

### What happened?

**_Probably caused by changes for #5660_**, the interface in **ETL metadata injection** to inject values in a **Merge rows (diff)** transform has changed: as a result, you'll get a warning about missing mappings to be removed.

This is what it looked like in Hop 2.15:
<img width="971" height="177" alt="Image" src="https://github.com/user-attachments/assets/354f704c-1b32-4a3b-9a1b-9d22ae3db636" />

This is what it looks like in Hop 2.16:
<img width="976" height="346" alt="Image" src="https://github.com/user-attachments/assets/608a80d5-6e82-4e49-8258-c84f06384522" />

Moreover, the basic log trace reports some errors like:
```
2025/11/21 02:01:22 - Leggi e confronta file.0 - ERROR: Target key 'VALUE_FIELDS' is not defined in '_loop-update-data-tables' pipeline!
2025/11/21 02:01:22 - Leggi e confronta file.0 - ERROR: Target key 'KEY_FIELDS' is not defined in '_loop-update-data-tables' pipeline!
```

However, it seems that pipelines developed with a previous version of Hop are still working as expected (this needs to be more accurately verified - I will do a comparison test tonight)

### Issue Priority

Priority: 2

### Issue Component

Component: Hop Gui, Component: Transforms, Component: Metadata

Update here... I can confirm with a sample case that something has broken between 2.15 and 2.16.

Need these files to test (place all the files in the `${PROJECT_HOME}` folder and change the `.hpl.txt` extension with just `.hpl`):
- [_test_merge_rows_injection.hpl.txt](https://github.com/user-attachments/files/23745859/_test_merge_rows_injection.hpl.txt): the "main" or "injecting" pipeline
- [_inj-merge-rows.hpl.txt](https://github.com/user-attachments/files/23745882/_inj-merge-rows.hpl.txt): the "injected" pipeline
- [sample1.txt](https://github.com/user-attachments/files/23745891/sample1.txt) and [sample2.txt](https://github.com/user-attachments/files/23745893/sample2.txt): sample data with 30 rows each, 10 are identical, 10 are changed, 10 are new/deleted 

When executing the main pipeline in Hop 2.15, the expected output is returned ([sample_compare_2.15.txt](https://github.com/user-attachments/files/23745962/sample_compare_2.15.txt)).

In Hop 2.16, instead, without changes in the main pipeline, "deleted" rows are stripped from the compare and all the other rows are considered identical ([sample_compare_2.16.txt](https://github.com/user-attachments/files/23745986/sample_compare_2.16.txt)).

I also tried to remove the missing mappings, and then move the key and value field in the new rows (see [_test_merge_rows_injection_update.hpl.txt](https://github.com/user-attachments/files/23746019/_test_merge_rows_injection_update.hpl.txt)), but I got the following error when executing the pipeline:
```
2025/11/25 12:35:09 - ETL metadata injection.0 - ERROR: Unexpected error
2025/11/25 12:35:09 - ETL metadata injection.0 - ERROR: org.apache.hop.core.exception.HopException: 
2025/11/25 12:35:09 - ETL metadata injection.0 - Error inject property 'KEY_FIELD' into class org.apache.hop.pipeline.transforms.mergerows.MergeRowsMeta
2025/11/25 12:35:09 - ETL metadata injection.0 - Cannot invoke "org.apache.hop.core.RowMetaAndData.getString(String, String)" because "data" is null
2025/11/25 12:35:09 - ETL metadata injection.0 - 
2025/11/25 12:35:09 - ETL metadata injection.0 - 	at org.apache.hop.core.injection.bean.BeanInjector.setProperty(BeanInjector.java:198)
2025/11/25 12:35:09 - ETL metadata injection.0 - 	at org.apache.hop.pipeline.transforms.metainject.MetaInject.newInjectionConstants(MetaInject.java:486)
2025/11/25 12:35:09 - ETL metadata injection.0 - 	at org.apache.hop.pipeline.transforms.metainject.MetaInject.processRow(MetaInject.java:135)
2025/11/25 12:35:09 - ETL metadata injection.0 - 	at org.apache.hop.pipeline.transform.RunThread.run(RunThread.java:54)
2025/11/25 12:35:09 - ETL metadata injection.0 - 	at java.base/java.lang.Thread.run(Thread.java:833)
2025/11/25 12:35:09 - ETL metadata injection.0 - Caused by: java.lang.NullPointerException: Cannot invoke "org.apache.hop.core.RowMetaAndData.getString(String, String)" because "data" is null
2025/11/25 12:35:09 - ETL metadata injection.0 - 	at org.apache.hop.core.injection.bean.BeanInjector.setProperty(BeanInjector.java:261)
2025/11/25 12:35:09 - ETL metadata injection.0 - 	at org.apache.hop.core.injection.bean.BeanInjector.setProperty(BeanInjector.java:190)
2025/11/25 12:35:09 - ETL metadata injection.0 - 	... 4 more
2025/11/25 12:35:09 - ETL metadata injection.0 - Finished processing (I=0, O=0, R=6, W=0, U=0, E=1)
```

Hope this helps for the resolution :)

--- a/integration-tests/mdi/0030-merge-rows.hpl
+++ b/integration-tests/mdi/0030-merge-rows.hpl
@@ -107,17 +107,10 @@ limitations under the License.
     <mappings>
       <mapping>
         <target_transform_name>Merge rows</target_transform_name>
-        <target_attribute_key>VALUE_FIELD</target_attribute_key>
-        <target_detail>Y</target_detail>
+        <target_attribute_key>VALUE_FIELDS</target_attribute_key>
+        <target_detail>N</target_detail>
         <source_transform>values</source_transform>
         <source_field>value</source_field>
-      </mapping>
-      <mapping>
-        <target_transform_name>Merge rows</target_transform_name>
-        <target_attribute_key>KEY_FIELD</target_attribute_key>
-        <target_detail>Y</target_detail>
-        <source_transform>keys</source_transform>
-        <source_field>key</source_field>
       </mapping>
       <mapping>
         <target_transform_name>Merge rows</target_transform_name>
@@ -128,17 +121,10 @@ limitations under the License.
       </mapping>
       <mapping>
         <target_transform_name>Merge rows</target_transform_name>
-        <target_attribute_key>COMPARE_TRANSFORM</target_attribute_key>
+        <target_attribute_key>KEY_FIELDS</target_attribute_key>
         <target_detail>N</target_detail>
-        <source_transform>metadata</source_transform>
-        <source_field>compare</source_field>
-      </mapping>
-      <mapping>
-        <target_transform_name>Merge rows</target_transform_name>
-        <target_attribute_key>REFERENCE_TRANSFORM</target_attribute_key>
-        <target_detail>N</target_detail>
-        <source_transform>metadata</source_transform>
-        <source_field>reference</source_field>
+        <source_transform>keys</source_transform>
+        <source_field>key</source_field>
       </mapping>
     </mappings>
     <attributes/>

--- a/plugins/transforms/mergerows/src/main/java/org/apache/hop/pipeline/transforms/mergerows/MergeRowsMeta.java
+++ b/plugins/transforms/mergerows/src/main/java/org/apache/hop/pipeline/transforms/mergerows/MergeRowsMeta.java
@@ -68,18 +68,16 @@ public class MergeRowsMeta extends BaseTransformMeta<MergeRows, MergeRowsData> {
   @HopMetadataProperty(
       key = "key",
       groupKey = "keys",
-      injectionKey = "KEY_FIELD",
+      injectionKey = "KEY_FIELDS",
       injectionKeyDescription = "MergeRows.Injection.KEY_FIELD",
-      injectionGroupKey = "KEY_FIELDS",
       injectionGroupDescription = "MergeRows.Injection.KEY_FIELDS")
   private List<String> keyFields;
 
   @HopMetadataProperty(
       key = "value",
       groupKey = "values",
-      injectionKey = "VALUE_FIELD",
+      injectionKey = "VALUE_FIELDS",
       injectionKeyDescription = "MergeRows.Injection.VALUE_FIELD",
-      injectionGroupKey = "VALUE_FIELDS",
       injectionGroupDescription = "MergeRows.Injection.VALUE_FIELDS")
   private List<String> valueFields;
 

